bpo-34206: Improve docs and test coverage for pre-init functions

ncoghlan opened the pull request:
- move the Py_Main documentation from the very high level API section
  to the initialization and finalization section
- make it clear that it encapsulates a full Py_Initialize/Finalize
  cycle of its own
- point out that exactly which settings will be read and applied
  correctly when called after a separate Py_Initialize or
  Py_InitalizeEx call is version dependent
- be explicit that Py_IsInitialized can be called prior to
  initialization
- actually test that Py_IsInitialized can be called prior to
  initialization
- move the docs for the new Py_BytesMain and Py_InitializeFromConfig
  functions to the initialization and finalization section
- ensure the new configuration and pre-initialization APIs are explicitly
  listed as being safe to call prior to interpreter initialization
- other minor tweaks to improve the integration between the docs for
  the traditional initialization API and the new struct-based one.

Closes #78387

<!-- issue-number: [bpo-34206](https://bugs.python.org/issue34206) -->
https://bugs.python.org/issue34206
<!-- /issue-number -->

ncoghlan committed:
bpo-34008: Py_Main and Py_IsInitialized are preinit functions

- move the Py_Main documentation from the very high level API section
  to the initialization and finalization section
- make it clear that it encapsulates a full Py_Initialize/Finalize
  cycle of its own
- point out that exactly which settings will be read and applied
  correctly when called after a separate Py_Initialize or
  Py_InitalizeEx call is version dependent
- be explicit that Py_IsInitialized can be called prior to
  initialization
- actually test that Py_IsInitialized can be called prior to
  initialization

ncoghlan committed:
Add news entry

ncoghlan committed:
Fix directive markup

ncoghlan committed:
Reword Py_Main docs

vstinner commented on Doc/c-api/veryhigh.rst:
```diff
@@ -24,22 +24,10 @@ use different libraries, so care should be taken that :c:type:`FILE\*` parameter
-   function will not return ``1``, but exit the process, as long as
-   ``Py_InspectFlag`` is not set.
+   The documentation for :c:func:`Py_Main` that previously appeared in this
+   section has been moved to :ref:`initialization`.
```
I don't think that this note is useful.

ncoghlan replied:
My rationale for including it is that any old links to https://docs.python.org/3/c-api/veryhigh.html#c.Py_Main will still end up here.

OTOH, https://www.google.com/search?q=link%3Ahttps%3A%2F%2Fdocs.python.org%2F3%2Fc-api%2Fveryhigh.html%23c.Py_Main strongly suggests that direct links to that are rather rare.

ncoghlan replied:
Removed

vstinner commented on Programs/_testembed.c:
```diff
@@ -149,6 +149,13 @@ static int test_pre_initialization_api(void)
+
+
     _Py_EMBED_PREINIT_CHECK("Initializing interpreter\n");
     Py_Initialize();
```
You can also test Py_IsInitialized() after Py_Initialize().

ncoghlan replied:
Done

ncoghlan committed:
Replace note with NEWS entry for Py_Main docs move

ncoghlan committed:
Further pre-initialization test improvements

- also check Py_IsInitialized while the interpreter is initialized
  and after it is finalized
- flush stdout in the embedding tests that run code so it appears
  in the expected order when running with "-vv"
- make "-vv" on the subinterpreter embedding tests less spammy

ncoghlan committed:
Adjust wording to be more accurate for 3.6 and 3.7

ncoghlan committed:
Add missing word

ncoghlan committed:
Fix typo

ncoghlan committed:
Update NEWS entry

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -19,6 +19,12 @@ a few functions and the :ref:`global configuration variables
+
+  * :c:func:`Py_Initialize`
+  * :c:func:`Py_InitializeEx`
+  * :c:func:`Py_Main`
```
I may be nice to also document Py_BytesMain().

ncoghlan replied:
Yeah, this was a super-old PR that I had actually forgotten about - it predates PEP 587 and the other improvements that landed for 3.8.

I'll look at adding the new APIs here when I fix the conflicts (as I assume the conflicts are with the docs for the new APIs).

That won't be tonight though - maybe Sunday.

ncoghlan replied:
Yep, the conflict in veryhigh.rst was indeed with Py_BytesMain - I've moved that over to the init section, and added it to the pre-init safe list.

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -307,6 +314,47 @@ Initializing and finalizing the interpreter
+   environmental and command line configuration settings will be updated is
+   version dependent (as it depends on which settings correctly support
+   being modified after they have already been set once in the first call to
+   ``Py_Initialize`` or ``Py_InitializeEx``).
```
Can you please try to document Py_RunMain()? It may help to explain how Py_Main() is implemented.

Py_Main() pseudo-code (without error handling) looks like:

```
PyConfig config;
PyConfig_InitPythonConfig(&config);
PyConfig_SetArgv(&config, argc, argv);
Py_InitializeFromConfig(&config);
PyConfig_Clear(&config);

Py_RunMain();
```

ncoghlan replied:
Done (I also moved the docs for `PyConfig_InitPythonConfig` and `Py_RunMain` into this page)

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34008-py-main-after-py-initialize

ncoghlan committed:
Tweak wording of Py_Main and Py_BytesMain docs

ncoghlan committed:
Tweak presentation order of init functions

ncoghlan committed:
Further integrate old and new init API docs

ncoghlan committed:
Merge in online PR changes, move InitalizeFromConfig to main init section

ncoghlan committed:
Merge branch 'master' into bpo-34008-py-main-after-py-initialize

ncoghlan committed:
Merge remote-tracking branch 'origin/main' into bpo-34008-py-main-after-py-initialize

ncoghlan committed:
Merge branch 'main' into bpo-34008-py-main-after-py-initialize

ncoghlan committed:
Fix merge error

ncoghlan committed:
Fix up NEWS entries

ncoghlan committed:
Actually reference the init-from-config anchor

ncoghlan commented on Doc/c-api/init.rst:
```diff
@@ -346,34 +361,42 @@ Initializing and finalizing the interpreter
+
+   See the :ref:`init-config` section for details on pre-initializing the
+   interpreter, populating the runtime configuration structure, and querying
+   the returned status structure.
```
I'd like to move the `Py_InitializeFromConfig` link target here so it appears directly after `Py_Initialize` and `Py_InitializeEx`. The full explanation of what it does depends on the rest of the `init-config` docs though, so those docs are cross-referenced for the details rather than attempting to put them inline.

ncoghlan committed:
Eliminate Py_RunMain docs duplication

ncoghlan committed:
Fix anchor syntax

ncoghlan commented on Doc/c-api/init.rst:
```diff
@@ -436,10 +459,102 @@ Initializing and finalizing the interpreter
 
-   This is a backwards-compatible version of :c:func:`Py_FinalizeEx` that
-   disregards the return value.
+.. c:function:: int Py_Main(int argc, wchar_t **argv)
```
Moving the `Py_Main` docs is where this change started (as described in the linked issue).

ncoghlan commented on Misc/NEWS.d/next/C API/2018-06-30-21-48-16.bpo-34008.2Wjtm0.rst:
```diff
@@ -0,0 +1,2 @@
+Added ``Py_IsInitialized`` to the list of APIs that are safe to call before
+the interpreter is initialized, and updated the embedding tests to cover it.
```
I didn't want to put this in the `Documentation` section, since the tests did change, and one entry in the C API sections seems neater than mentioning it as both a docs and test suite change.

ncoghlan commented on Doc/c-api/init.rst:
```diff
@@ -436,10 +459,102 @@ Initializing and finalizing the interpreter
+.. c:function:: int Py_Main(int argc, wchar_t **argv)
+
+   The main program for the standard interpreter, encapsulating a full
+   :c:func:`Py_Initialize`/:c:func:`Py_Finalize` cycle, as well as additional
```
Fix stale reference to the deprecated `Py_Finalize` function:

```suggestion
   initialization/finalization cycle, as well as additional
```

ncoghlan committed:
Replace stale reference to Py_Finalize

ncoghlan commented on Doc/c-api/init.rst:
```diff
@@ -436,10 +459,102 @@ Initializing and finalizing the interpreter
+   any other unhandled exception.
+
+   This function always finalizes the Python interpreter regardless of whether
+   it returns a value or immediately exits the process due to an an unhandled
```
```suggestion
   it returns a value or immediately exits the process due to an unhandled
```

ncoghlan committed:
Fix typo

ncoghlan commented on Doc/c-api/init.rst:
```diff
@@ -346,34 +361,42 @@ Initializing and finalizing the interpreter
-
-   Use the :c:func:`Py_InitializeFromConfig` function to customize the
+   This initializes the table of loaded modules (``sys.modules``), and creates
+   the fundamental modules :mod:`builtins`, :mod:`__main__` and :mod:`sys`. 
```
```suggestion
   the fundamental modules :mod:`builtins`, :mod:`__main__` and :mod:`sys`.
```

ncoghlan committed:
Remove trailing whitespace

ncoghlan committed:
Merge branch 'main' into bpo-34008-py-main-after-py-initialize

ncoghlan committed:
Add back Py_Finalize docs (presumably lost in a merge error)

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -21,6 +22,15 @@ a few functions and the :ref:`global configuration variables
+
+  * :c:func:`Py_Initialize`
+  * :c:func:`Py_InitializeEx`
+  * :c:func:`Py_InitializeFromConfig`
```
There is now also Py_InitializeFromInitConfig() (PEP 741).

ncoghlan replied:
I'll leave it out for now since I'm going to backport this to the 3.13 docs, but we should add it for 3.14.

willingc commented on Misc/NEWS.d/next/Documentation/2018-07-04-20-35-25.bpo-34008.bqecIb.rst:
```diff
@@ -0,0 +1,8 @@
+The ``Py_Main`` documentation moved from the "Very High Level API" section to the
```
```suggestion
The :c:func:`Py_Main` documentation moved from the "Very High Level API" section to the
```

willingc commented on Misc/NEWS.d/next/Documentation/2018-07-04-20-35-25.bpo-34008.bqecIb.rst:
```diff
@@ -0,0 +1,8 @@
+of ``Py_Initialize`` rather than after it (since ``Py_Main`` makes its own
+call to ``Py_Initialize``). However, also document that calling both is
+supported, but it's version dependent as to exactly which settings
+will be applied correctly in that case.
```
```suggestion
call to ``Py_Initialize``). Document that calling both is
supported but is version dependent on which settings
will be applied correctly.
```

willingc reviewed (approved):
Docs look good to me. Thanks @ncoghlan.

ncoghlan committed:
Apply suggestions from code review

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>
